### PR TITLE
CDAP-531: CLI (common) - Need to have inbuilt Help and Exit commands

### PR DESCRIPTION
**Pull request by ghost**
These changes do not affect the current COOPR, CDAP CLIs, but need to remove Exit and Help commands from other CLIs. I will do this after this PR will be merged. Here also added QuitCommand due to https://issues.cask.co/browse/COOPR-623. @wolf31o2 please take a look. I have tested this changes on COOPR CLI.

### CODE REVIEW

File: common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        for (Command command : commands) {
+          printStream.println(String.format("%s: %s", command.getPattern(), command.getDescription()));
+        }
+      }
+    } else {
+      if (helpHeader != null) {
+        printStream.println(helpHeader);
+        printStream.println();
+      }
+      printStream.println(String.format("Available commands: \n%s: %s", getPattern(), getDescription()));
```
Comment (gokulavasan) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: I guess we don't need to print getPattern(), getDescription() here.

File: common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  @Override
+  public String getPattern() {
+    return String.format("help [<%s>]", COMMAND_KEY);
+  }
+
+  @Override
+  public String getDescription() {
+    return "Prints this helper text";
```
Comment (gokulavasan) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Prints usage information and description of commands

File: common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java (hunk 1 above)
Comment (ghost) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Then we will not print description of help command.

Comment (gokulavasan) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Ok. Didn't notice the newline in string. Ignore my comment.

Comment (ghost) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Ok. Thanks

Comment (gokulavasan) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Oh, will 
`> help help`
work?

File: common-cli/src/main/java/co/cask/common/cli/command/QuitCommand.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.common.cli.command;
+
+/**
+ * Quits the cli.
```
Comment (awholegunch) [inline common-cli/src/main/java/co/cask/common/cli/command/QuitCommand.java]: Quits the command-line interface.

File: common-cli/src/main/java/co/cask/common/cli/command/QuitCommand.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+public class QuitCommand extends AbstractExitCommand {
+
+  @Override
+  public String getPattern() {
+    return "quit";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Quits the cli";
```
Comment (awholegunch) [inline common-cli/src/main/java/co/cask/common/cli/command/QuitCommand.java]: Quits the command-line interface

File: common-cli/src/main/java/co/cask/common/cli/command/ExitCommand.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.common.cli.command;
+
+/**
+ * Exits the cli.
```
Comment (awholegunch) [inline common-cli/src/main/java/co/cask/common/cli/command/ExitCommand.java]: Exits the command-line interface.

File: common-cli/src/main/java/co/cask/common/cli/command/ExitCommand.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+public class ExitCommand extends AbstractExitCommand {
+
+  @Override
+  public String getPattern() {
+    return "exit";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Exits the cli";
```
Comment (awholegunch) [inline common-cli/src/main/java/co/cask/common/cli/command/ExitCommand.java]: Exits the command-line interface

File: common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java (hunk 1 above)
Comment (ghost) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: No. I will update PR to fix this.

File: common-cli/src/main/java/co/cask/common/cli/CommandSet.java [hunk 7]
```diff
@@ -73,6 +74,22 @@ public CommandMatch findMatch(String input) throws InvalidCommandException {
     throw new InvalidCommandException(input);
   }
 
+  /**
+   * Finds a matching commands for the provided input.
```
Comment (awholegunch) [inline common-cli/src/main/java/co/cask/common/cli/CommandSet.java]: Finds matching commands for the provided input.

File: common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java (hunk 1 above)
Comment (ghost) [inline common-cli/src/main/java/co/cask/common/cli/command/HelpCommand.java]: Works for now

**Comment by ghost** [comment]
Updated PR. Fix this https://github.com/caskdata/common/pull/13#discussion_r21110520

**Comment by gokulavasan** [comment]
LGTM :+1:

**Comment by awholegunch** [comment]
LGTM.

**Comment by gokulavasan** [comment]
@ytalashkocv Have you tested the changes with Coopr CLI after the PR comments?

**Comment by gokulavasan** [comment]
@ytalashkocv Can you add unit tests for these changes?

**Comment by ghost** [comment]
@gokulavasan yes I have tested the changes with Coopr CLI after the PR comments

**Comment by ghost** [comment]
@gokulavasan Ok, I will add unit tests

**Comment by jwang47** [comment]
Let's merge after unit tests are added.

**Comment by ghost** [comment]
Added unit tests

**Comment by jwang47** [comment]
LGTM

**Pull request merged**